python/ray/serve/task_processor.py

```diff
@@ -0,0 +1,248 @@
+import logging
+import threading
+import time
+from abc import ABC, abstractmethod
+from typing import Any, Dict, Optional
+from unittest.mock import MagicMock
+
+from celery import Celery
+
+from ray.serve._private.constants import SERVE_LOGGER_NAME
+from ray.serve.schema import (
+    CeleryTaskProcessorConfig,
+    MockTaskProcessorConfig,
+    TaskProcessorConfig,
+    TaskResult,
+)
+
+logger = logging.getLogger(SERVE_LOGGER_NAME)
+
+
+class TaskProcessorAdapter(ABC):
+    @abstractmethod
+    def initialize(self, config: TaskProcessorConfig):
+        pass
+
+    @abstractmethod
+    def register_task_handle(self, func, name=None):
+        pass
+
+    @abstractmethod
+    def enqueue_task(self, task_name, args=None, kwargs=None, **options) -> TaskResult:
+        pass
+
+    @abstractmethod
+    def get_task_status(self, task_id) -> Dict[str, Any]:
+        pass
+
+    @abstractmethod
+    def cancel_task(self, task_id) -> bool:
+        pass
+
+    @abstractmethod
+    def get_metrics(self) -> Dict[str, Any]:
+        pass
+
+    @abstractmethod
+    def start_consumer(self, **kwargs):
+        pass
+
+    @abstractmethod
+    def stop_consumer(self, timeout: float = 10.0):
+        pass
+
+    @abstractmethod
+    def shutdown(self):
+        pass
+
+    @abstractmethod
+    def health_check(self):
+        pass
+
+
+class CeleryTaskProcessorAdapter(TaskProcessorAdapter):
+    _app: Celery
+    _config: TaskProcessorConfig
+    _worker_thread: Optional[threading.Thread] = None
```

Comment:
Curious why you are using the singleton pattern here, what's the downside to defining these in the class constructor?